Csc task crash during build cancelation

**Version Used**: 

Whatever is in 6.0.101 SDK.

**Steps to Reproduce**:

1. build and hit cancel at the perfect moment (it's probably very rare to repro)

**Expected Behavior**:

Cancels cleanly

**Actual Behavior**:

It seems similar to https://github.com/dotnet/roslyn/issues/31950, but a slightly different place. 

One option is to catch the ODE but that would likely be papering over some mis-synchronization here.


```
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018: The "Csc" task failed unexpectedly. [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Cannot access a closed pipe.)) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:  ---> System.AggregateException: One or more errors occurred. (Cannot access a closed pipe.) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:  ---> System.ObjectDisposedException: Cannot access a closed pipe. [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.IO.Pipes.PipeStream.get_SafePipeHandle() [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.IO.Pipes.PipeStream.PipeValueTaskSource.<>c.<RegisterForCancellation>b__17_0(Object s, CancellationToken token) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.Invoke(Delegate d, Object state, CancellationTokenSource source) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.CallbackNode.ExecuteCallback() [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    --- End of inner exception stack trace --- [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.NotifyCancellation(Boolean throwOnFirstException) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.LinkedNCancellationTokenSource.<>c.<.cctor>b__4_0(Object s) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.Invoke(Delegate d, Object state, CancellationTokenSource source) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.CallbackNode.ExecuteCallback() [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    --- End of inner exception stack trace --- [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.NotifyCancellation(Boolean throwOnFirstException) [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at System.Threading.CancellationTokenSource.Cancel() [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at Microsoft.CodeAnalysis.BuildTasks.ManagedCompiler.Cancel() [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
C:\Program Files\dotnet\sdk\6.0.101\Roslyn\Microsoft.CSharp.Core.targets(75,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Cancel() [C:\git\runtime\src\libraries\System.Runtime.Numerics\tests\System.Runtime.Numerics.Tests.csproj]
    0 Warning(s)
    1 Error(s)
```

--- a/src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.ValueTaskSource.cs
+++ b/src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.ValueTaskSource.cs
@@ -75,7 +75,7 @@ namespace System.IO.Pipes
                         _cancellationRegistration = cancellationToken.UnsafeRegister(static (s, token) =>
                         {
                             PipeValueTaskSource vts = (PipeValueTaskSource)s!;
-                            if (vts._pipeStream.InternalHandle is { IsInvalid: false } handle)
+                            if (vts._pipeStream.InternalHandle is { IsClosed: false } handle)
                             {
                                 try
                                 {

--- a/src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.ValueTaskSource.cs
+++ b/src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.ValueTaskSource.cs
@@ -75,11 +75,11 @@ namespace System.IO.Pipes
                         _cancellationRegistration = cancellationToken.UnsafeRegister(static (s, token) =>
                         {
                             PipeValueTaskSource vts = (PipeValueTaskSource)s!;
-                            if (!vts._pipeStream.SafePipeHandle.IsInvalid)
+                            if (vts._pipeStream.InternalHandle is { IsInvalid: false } handle)
                             {
                                 try
                                 {
-                                    Interop.Kernel32.CancelIoEx(vts._pipeStream.SafePipeHandle, vts._overlapped);
+                                    Interop.Kernel32.CancelIoEx(handle, vts._overlapped);
                                     // Ignore all failures: no matter whether it succeeds or fails, completion is handled via the IOCallback.
                                 }
                                 catch (ObjectDisposedException) { } // in case the SafeHandle is (erroneously) closed concurrently

